cylc/flow/network/graphql.py

```diff
@@ -0,0 +1,372 @@
+                continue
+            field_def = get_field_def(
+                self.schema, parent_type, field.name.value)
+            if field_def is None:
+                continue
+            arg_vals = get_argument_values(
+                field_def.args, field.arguments, variables)
+            if arg_vals.get(arg_name) == arg_value:
+                return True
+            if field.selection_set is None:
+                continue
+            if self.args_selection_search(
+                    field.selection_set, variables,
+                    get_named_type(field_def.type), arg_name, arg_value):
+                return True
+        return False
+
+
+def execute_and_validate(
+        schema,  # type: GraphQLSchema
+        document_ast,  # type: Document
+        *args,  # type: Any
+        **kwargs  # type: Any
+):
+    # type: (...) -> Union[ExecutionResult, Observable]
+    """Validate schema, and execute request doc against it."""
+    do_validation = kwargs.get("validate", True)
+    if do_validation:
+        validation_errors = validate(schema, document_ast)
+        if validation_errors:
+            return ExecutionResult(errors=validation_errors, invalid=True)
+
+    result = execute(schema, document_ast, *args, **kwargs)
+
+    # Search request docuement to determine if 'stripNull: true' is set
+    # as and argument. It can not be done in the middleware, as they
+    # can be Promises/futures (so may not been resolved at this point).
+    variable_values = kwargs['variable_values'] or {}
+    doc_args = AstDocArguments(schema, document_ast, variable_values)
+    if doc_args.has_arg_val(STRIP_ARG, True):
+        if kwargs.get('return_promise', False):
+            return result.then(null_stripper)
+        return null_stripper(result)
+    return result
+
+
+class GraphQLCoreBackend(GraphQLBackend):
+    """GraphQLCoreBackend will return a document using the default
+    graphql executor"""
+
+    def __init__(self, executor=None):
+        # type: (Optional[Any]) -> None
+        self.execute_params = {"executor": executor}
+
+    def document_from_string(self, schema, document_string):
+        # type: (GraphQLSchema, Union[Document, str]) -> GraphQLDocument
+        """Parse string and setup request docutment for execution."""
+        if isinstance(document_string, ast.Document):
+            document_ast = document_string
+            document_string = print_ast(document_ast)
+        else:
+            if not isinstance(document_string, str):
+                logger.error("The query must be a string")
+            document_ast = parse(document_string)
+        return GraphQLDocument(
+            schema=schema,
+            document_string=document_string,
+            document_ast=document_ast,
+            execute=partial(
+                execute_and_validate,
+                schema,
+                document_ast,
+                **self.execute_params
+            ),
+        )
```

Comment:
This block looks identical to the original, can we just import it to avoid duplication.

Reply:
No, we would need to subclass it and overwrite `document_from_string`, because it calls `execute_and_validate` which we have changed.

But considering that it is already doing that with `GraphQLBackend` I didn't see the point... And it may change with upgrade from legacy.

Reply:
As mentioned below, wrapped the `execute_and_validate` in a `execute_and_validate_and_strip`... So altered this class, with both this callable and a little documentation of the `stripNull` used...

Reply:
Ok, short of monkey patching I can't see how we could achieve this by import so fair enough.

Licensing is not my strong suit, thankfully its MIT so shouldn't be an issue, however I think we may need to do one or more of the below:

* Copy in file headers from their project.
* Add an acknowledgement (with/in this block of code).
* Put this class in a separate file with their copyright header rather than ours.

@hjoliver any ideas. I think if we keep this code away from our copyright headers we should need to run this through legal?

Reply:
@oliver-sanders - Given that there is no function and/or class exactly the same in `cylc.flow.network.graphql`, I'm struggling to see why we need to bother.

It's true that our `GraphQLCoreBackend` needs to subclass `GraphQLBackend` and has one similar block in the method in the original.. But it's necessarily different because of the script it calls... Should I name it `GraphQLCylcBackend`? (what does it have to be 33% different? it's such a simple subclassing, and obvious that we depend on `graphql-core`)

Maybe we could just get the `GraphQLDocument` result and modify the `execute` attribute, because other libraries depend on it (i.e. create another class subclassing this one with the same method calling super)...

Anyway, kinda pointless, because things might change when I revisit with change to `graphql-core 3+`... And again no item in our module is an exact copy..

Anyway, open to suggestions, (@hjoliver? how should I approach this sort of thing).

Reply:
> I'm struggling to see why we need to bother.

This code is a direct copy, even the comments are copied. **If** there are any licensing issues with putting this under the Cylc project copyright notice (which there might not be) we need to address them pre-merge whether there is an intention to change it later or not.

The MIT license requires the copyright notice to be preserved (caveat https://opensource.stackexchange.com/a/2188).

Reply:
> This code is a direct copy, even the comments are copied.

Wrong. The closest thing to a copy in `cylc.flow.network.graphql` is the `GraphQLCoreBackend` class, which has now been modified slightly for Cylc.  [I think] everything else is completely Cylc.

Reply:
Check again.

Reply:
So again, and otherwise, how much modification is need to forgo the label "copy"...

Creating a class, sub-classing another, then overwriting a method doesn't seem a lot different to creating a module importing another module and overwriting some classes to act different.

Reply:
Ok further modifications (i.e. not a copy), and credited origin:
https://github.com/dwsutherland/cylc-flow/blob/92fdab2e61dab5244f77d8c57b742d75a3000a01/cylc/flow/network/graphql.py#L249-L301

Likened to a paraphrased story with a different twist at the end..
i.e. The boat is the same, but the journey is different, so thank you `graphql-core` for the boat.

Reply:
Licensing is a pain, I think that code is now sufficiently different, it would be hard to implement it in any other way.